### PR TITLE
More details of error of getting op matched sbp signature

### PR DESCRIPTION
**Pull request by leaves-zwx**
丰富在 2D SBP 推导过程中找不到匹配的 signature 的报错信息，以便于判断问题。

报错效果：

```
F1222 10:01:55.906486 146584 op_graph.cpp:313]
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/graph/op_graph.cpp", line 313, in InferOpNodeNdSbpSignature
    op_node->mut_op()->InferNdSbpSignatureIf(nd_sbp_sig_conf, op_node->parallel_desc(), NdSbpInferHint4Ibn)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/operator/operator.cpp", line 664, in InferNdSbpSignatureIf
    InferNdSbpSignature(&nd_sbp_signature, nd_sbp_constraints, parallel_desc, NdSbpInferHint4Ibn)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/operator/user_op.cpp", line 757, in InferNdSbpSignature
    Operator::InferNdSbpSignature(nd_sbp_signature, nd_sbp_constraints, parallel_desc, NdSbpInferHint4Ibn)
op: model.classifier-matmul_837_grad_b can't find available sbp signature.
Supported SBP signatures are: (b_0, a_0) -> (out_0): [
        (B, S(1)) -> (S(0)),
        (S(1), B) -> (S(1)),
        (S(0), S(0)) -> (P),
        (B, P) -> (P),
        (P, B) -> (P),
        (B, B) -> (B),
], but got (B, S(0)) -> ? at hierarchy dim 0, since the SBP of inputs are: b_0: (B, B), a_0: (S(0), S(0))
*** Check failure stack trace: ***
```

```
oneflow._oneflow_internal.exception.RuntimeException:
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/functional/impl/nn_functor.cpp", line 888, in LazySparseSoftmaxCrossEntropyMsOperator
    OpInterpUtil::Dispatch<TensorTuple>( *op_sparse_softmax_cross_entropy_ms_, {logits, label}, attrs)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/framework/op_interpreter/op_interpreter_util.cpp", line 131, in Dispatch<oneflow::one::TensorTuple>
    Dispatch(op_expr, inputs, outputs.get(), ctx)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/framework/op_interpreter/op_interpreter.cpp", line 95, in Apply
    internal_->Apply(op_expr, inputs, outputs, ctx)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp", line 655, in ApplyImpl
    infer_ctx->AddAndInferConsistentOp(*op_conf)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/job/job_build_and_infer_ctx.cpp", line 599, in AddAndInferOp
    InferOpOutNdSbp(op, nd_sbp_sig_conf, parallel_desc)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/job/job_build_and_infer_ctx.cpp", line 255, in InferOpOutNdSbp
    op->InferNdSbpSignatureIf(nd_sbp_sig_conf, parallel_desc, NdSbpInferHint4Ibn)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/operator/operator.cpp", line 664, in InferNdSbpSignatureIf
    InferNdSbpSignature(&nd_sbp_signature, nd_sbp_constraints, parallel_desc, NdSbpInferHint4Ibn)
  File "/home/zhangwenxiao/repos/oneflow/oneflow/core/operator/user_op.cpp", line 757, in InferNdSbpSignature
    Operator::InferNdSbpSignature(nd_sbp_signature, nd_sbp_constraints, parallel_desc, NdSbpInferHint4Ibn)
op: loss_fn-sparse_softmax_cross_entropy_ms_840 can't find available sbp signature.
Supported SBP signatures are: (label_0, prediction_0) -> (prob_0, out_0): [
        (S(0), S(0)) -> (S(0), S(0)),
        (B, S(1)) -> (S(1), P),
        (B, B) -> (B, B),
], but got (S(0), B) -> ? at hierarchy dim 0, since the SBP of inputs are: label_0: (S(0), S(0)), prediction_0: (B, S(1))
```

### CODE REVIEW

File: oneflow/core/job/sbp_parallel.cpp [hunk 1]
```diff
@@ -227,4 +227,57 @@ void CheckSbpSignatureAndNdSbpEquals(const cfg::SbpSignature& sbp_sig,
   }
 }
 
+Maybe<std::string> StringifySbpSignatureList(const PbRpf<std::string>& inputs,
```
Comment (hjchen2) [inline oneflow/core/job/sbp_parallel.cpp:230]: 一般会命名为SbpSignatureListAsString，或者SerializeSbpSignatureListToString

File: oneflow/core/job/sbp_parallel.cpp [hunk 2]
```diff
@@ -227,4 +227,57 @@ void CheckSbpSignatureAndNdSbpEquals(const cfg::SbpSignature& sbp_sig,
...
+      }
+      return SbpParallelToString(it->second);
+    }));
+    ss << ",\n";
+  }
+  ss << "]";
+  return ss.str();
+}
+
+std::string StringifyNdSbp(const cfg::NdSbp& nd_sbp) {
```
Comment (hjchen2) [inline oneflow/core/job/sbp_parallel.cpp:272]: `core/framework/nd_sbp.cpp`里面已经有一个NdSbpToString，你看看是不是和你要实现的是一样的

File: oneflow/core/operator/operator.cpp [hunk 3]
```diff
@@ -738,7 +738,20 @@ Maybe<void> Operator::InferNdSbpSignature(
           break;
         }
       }
-      CHECK_OR_RETURN(matched_sbp_signature != nullptr) << " op_name " << op_name();
+      if (!matched_sbp_signature) {
+        std::ostringstream err;
+        err << "op: " << op_name() << " can't find available sbp signature at hierarchy dim " << i
+            << ".\nSupported SBP signatures are: ";
+        err << *JUST(StringifySbpSignatureList(input_bns(), output_bns(), list));
+        err << ", but got SBP of inputs are: ";
```
Comment (hjchen2) [inline oneflow/core/operator/operator.cpp:746]: 这里要不要把hierarchy dim i的sbp也打出来，比如报错信息这样写，"but got B, S(0) at hierarchy dim 0 since the SBP of inputs is (b_0: [B, B], a_0: [S(0), S(0)]) ."

Comment (leaves-zwx) [reply oneflow/core/operator/operator.cpp:746]: OK

File: oneflow/core/job/sbp_parallel.cpp (hunk 2 above)
Comment (leaves-zwx) [reply oneflow/core/job/sbp_parallel.cpp:272]: 确实有，我还是用之前提供的函数吧

Comment (leaves-zwx) [reply oneflow/core/job/sbp_parallel.cpp:272]: 我试了下，这个函数不是我想要的效果

```
op: model.classifier-matmul_837_grad_b can't find available sbp signature at hierarchy dim 0.
Supported SBP signatures are: (b_0, a_0) -> (out_0): [
        (oneflow.sbp.broadcast, oneflow.sbp.split(axis=1)) -> (oneflow.sbp.split(axis=0)),
        (oneflow.sbp.split(axis=1), oneflow.sbp.broadcast) -> (oneflow.sbp.split(axis=1)),
        (oneflow.sbp.split(axis=0), oneflow.sbp.split(axis=0)) -> (oneflow.sbp.partial_sum),
        (oneflow.sbp.broadcast, oneflow.sbp.partial_sum) -> (oneflow.sbp.partial_sum),
        (oneflow.sbp.partial_sum, oneflow.sbp.broadcast) -> (oneflow.sbp.partial_sum),
        (oneflow.sbp.broadcast, oneflow.sbp.broadcast) -> (oneflow.sbp.broadcast),
], but got SBP of inputs are: b_0: (oneflow.sbp.broadcast, oneflow.sbp.broadcast), a_0: (oneflow.sbp.split(axis=0), oneflow.sbp.split(axis=0))
```

我觉得这种表示不直观，跟直接打印 protobuf 没啥区别

Comment (leaves-zwx) [reply oneflow/core/job/sbp_parallel.cpp:272]: 我想要的是一种外部视角的表示，而不是内部视角的结构的表示

Comment (hjchen2) [reply oneflow/core/job/sbp_parallel.cpp:272]: 你写的NdSbpToString其实是原NdSbpToString的输出SBP简称版，我觉得可以放到一个文件下，只是命名上区分一下。
或者这么说，sbp其实都有简称和全称，比如oneflow.sbp.split(axis=0)是全称，S(0)是简称，在ToString时指定要全称还是简称，这种方式比两套接口我觉得好一些

File: oneflow/core/job/sbp_parallel.cpp (hunk 1 above)
Comment (leaves-zwx) [reply oneflow/core/job/sbp_parallel.cpp]: 已改

**Review by hjchen2 — APPROVED** [review]

**Review by wyg1997 — APPROVED** [review]

**Review by guo-ran — APPROVED** [review]

**Comment by hjchen2** [comment]
https://github.com/Oneflow-Inc/OneTeam/issues/883
